File: crates/codeprysm-core/tests/fixtures/integration_repos/go/sample.go
// Package sample provides integration testing fixtures for Go.
//
// This module demonstrates various Go language features for
// graph generation validation including structs, interfaces, and methods.
package sample

import (
	"context"
	"errors"
	"sync"
	"time"
)

// MaxItems is a module-level constant.
const MaxItems = 100

// ErrInvalidOperation is returned when an operation fails.
var ErrInvalidOperation = errors.New("invalid operation")

// UserRole represents the role of a user.
type UserRole string

const (
	RoleAdmin UserRole = "admin"
	RoleUser  UserRole = "user"
	RoleGuest UserRole = "guest"
)

// User represents a user entity.
type User struct {
	ID        string
	Name      string
	Email     string
	Role      UserRole
	CreatedAt time.Time
}

// Calculator is a simple calculator interface.
type Calculator interface {
	Add(amount int) int
	Multiply(factor int) int
	Value() int
}

// Repository is a generic repository interface.
type Repository[T any] interface {
	FindById(ctx context.Context, id string) (T, error)
	FindAll(ctx context.Context) ([]T, error)
	Save(ctx context.Context, item T) error
	Delete(ctx context.Context, id string) error
}

// SimpleCalculator implements the Calculator interface.
type SimpleCalculator struct {
	value   int
	history []int
	mu      sync.Mutex
}

// NewSimpleCalculator creates a new SimpleCalculator.
func NewSimpleCalculator(initialValue int) *SimpleCalculator {
	return &SimpleCalculator{
		value:   initialValue,
		history: make([]int, 0),
	}
}

// Add adds an amount to the current value.
func (c *SimpleCalculator) Add(amount int) int {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value += amount
	c.history = append(c.history, amount)
	return c.value
}

// Multiply multiplies the current value by a factor.
func (c *SimpleCalculator) Multiply(factor int) int {
	c.mu.Lock()
	defer c.mu.Unlock()
	c.value *= factor
	return c.value
}

// Value returns the current value.
func (c *SimpleCalculator) Value() int {
	c.mu.Lock()
	defer c.mu.Unlock()
	return c.value
}

// History returns a copy of the operation history.
func (c *SimpleCalculator) History() []int {
	c.mu.Lock()
	defer c.mu.Unlock()
	result := make([]int, len(c.history))
	copy(result, c.history)
	return result
}

// AsyncProcessor processes items asynchronously.
type AsyncProcessor struct {
	name           string
	processedCount int
	mu             sync.Mutex
}

// NewAsyncProcessor creates a new AsyncProcessor.
func NewAsyncProcessor(name string) *AsyncProcessor {
	return &AsyncProcessor{
		name:           name,
		processedCount: 0,
	}
}

// ProcessItem processes a single item.
func (p *AsyncProcessor) ProcessItem(ctx context.Context, item string) (string, error) {
	select {
	case <-ctx.Done():
		return "", ctx.Err()
	case <-time.After(10 * time.Millisecond):
		p.mu.Lock()
		p.processedCount++
		p.mu.Unlock()
		return p.name + ":" + item, nil
	}
}

// ProcessBatch processes multiple items.
func (p *AsyncProcessor) ProcessBatch(ctx context.Context, items []string) ([]string, error) {
	results := make([]string, 0, len(items))
	for _, item := range items {
		result, err := p.ProcessItem(ctx, item)
		if err != nil {
			return nil, err
		}
		results = append(results, result)
	}
	return results, nil
}

// ProcessedCount returns the number of processed items.
func (p *AsyncProcessor) ProcessedCount() int {
	p.mu.Lock()
	defer p.mu.Unlock()
	return p.processedCount
}

// DataProcessor is a generic data processor.
type DataProcessor[T any] struct {
	data []T
}

// NewDataProcessor creates a new DataProcessor.
func NewDataProcessor[T any]() *DataProcessor[T] {
	return &DataProcessor[T]{
		data: make([]T, 0),
	}
}

// Add adds an item to the processor.
func (d *DataProcessor[T]) Add(item T) {
	d.data = append(d.data, item)
}

// Map applies a function to all items.
func Map[T, U any](d *DataProcessor[T], fn func(T) U) []U {
	results := make([]U, len(d.data))
	for i, item := range d.data {
		results[i] = fn(item)
	}
	return results
}

// StandaloneFunction is a function outside any type.
func StandaloneFunction(param string) int {
	return len(param)
}

// AsyncStandalone is an async standalone function.
func AsyncStandalone(ctx context.Context, url string) (map[string]interface{}, error) {
	select {
	case <-ctx.Done():
		return nil, ctx.Err()
	case <-time.After(100 * time.Millisecond):
		return map[string]interface{}{"url": url}, nil
	}
}

// Square is a simple helper function.
func Square(x int) int {
	return x * x
}

File: crates/codeprysm-core/tests/fixtures/integration_repos/rust/sample.rs
//! Sample Rust module for integration testing.
//!
//! This module demonstrates various Rust language features for
//! graph generation validation including structs, traits, impls, and macros.

use std::collections::HashMap;
use std::sync::Mutex;

/// Module-level constant.
pub const MAX_ITEMS: usize = 100;

/// User role enum.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum UserRole {
    Admin,
    User,
    Guest,
}

/// User struct.
#[derive(Debug, Clone)]
pub struct User {
    pub id: String,
    pub name: String,
    pub email: String,
    pub role: UserRole,
}

/// Calculator trait defining calculator operations.
pub trait Calculator {
    fn add(&mut self, amount: i32) -> i32;
    fn multiply(&mut self, factor: i32) -> i32;
    fn value(&self) -> i32;
}

/// Repository trait for generic data access.
pub trait Repository<T> {
    fn find_by_id(&self, id: &str) -> Option<&T>;
    fn find_all(&self) -> Vec<&T>;
    fn save(&mut self, item: T) -> Result<(), &'static str>;
    fn delete(&mut self, id: &str) -> bool;
}

/// Simple calculator implementation.
pub struct SimpleCalculator {
    value: i32,
    history: Vec<i32>,
}

impl SimpleCalculator {
    /// Create a new calculator with an initial value.
    pub fn new(initial_value: i32) -> Self {
        Self {
            value: initial_value,
            history: Vec::new(),
        }
    }

    /// Get the operation history.
    pub fn history(&self) -> &[i32] {
        &self.history
    }

    /// Static method to square a number.
    pub fn square(x: i32) -> i32 {
        x * x
    }
}

impl Calculator for SimpleCalculator {
    fn add(&mut self, amount: i32) -> i32 {
        self.value += amount;
        self.history.push(amount);
        self.value
    }

    fn multiply(&mut self, factor: i32) -> i32 {
        self.value *= factor;
        self.value
    }

    fn value(&self) -> i32 {
        self.value
    }
}

/// Async processor for handling items.
pub struct AsyncProcessor {
    name: String,
    processed_count: Mutex<usize>,
}

impl AsyncProcessor {
    /// Create a new async processor.
    pub fn new(name: impl Into<String>) -> Self {
        Self {
            name: name.into(),
            processed_count: Mutex::new(0),
        }
    }

    /// Process a single item.
    pub async fn process_item(&self, item: &str) -> String {
        tokio::time::sleep(std::time::Duration::from_millis(10)).await;
        let mut count = self.processed_count.lock().unwrap();
        *count += 1;
        format!("{}:{}", self.name, item)
    }

    /// Process multiple items.
    pub async fn process_batch(&self, items: &[&str]) -> Vec<String> {
        let mut results = Vec::with_capacity(items.len());
        for item in items {
            let result = self.process_item(item).await;
            results.push(result);
        }
        results
    }

    /// Get the processed count.
    pub fn processed_count(&self) -> usize {
        *self.processed_count.lock().unwrap()
    }
}

/// Generic data processor.
pub struct DataProcessor<T> {
    data: Vec<T>,
}

impl<T> DataProcessor<T> {
    /// Create a new data processor.
    pub fn new() -> Self {
        Self { data: Vec::new() }
    }

    /// Add an item.
    pub fn add(&mut self, item: T) {
        self.data.push(item);
    }

    /// Map over all items.
    pub fn map<U, F>(&self, f: F) -> Vec<U>
    where
        F: Fn(&T) -> U,
    {
        self.data.iter().map(f).collect()
    }

    /// Filter items.
    pub fn filter<F>(&self, predicate: F) -> Vec<&T>
    where
        F: Fn(&T) -> bool,
    {
        self.data.iter().filter(|item| predicate(item)).collect()
    }
}

impl<T> Default for DataProcessor<T> {
    fn default() -> Self {
        Self::new()
    }
}

/// User repository implementation.
pub struct UserRepository {
    users: HashMap<String, User>,
}

impl UserRepository {
    /// Create a new user repository.
    pub fn new() -> Self {
        Self {
            users: HashMap::new(),
        }
    }
}

impl Default for UserRepository {
    fn default() -> Self {
        Self::new()
    }
}

impl Repository<User> for UserRepository {
    fn find_by_id(&self, id: &str) -> Option<&User> {
        self.users.get(id)
    }

    fn find_all(&self) -> Vec<&User> {
        self.users.values().collect()
    }

    fn save(&mut self, user: User) -> Result<(), &'static str> {
        self.users.insert(user.id.clone(), user);
        Ok(())
    }

    fn delete(&mut self, id: &str) -> bool {
        self.users.remove(id).is_some()
    }
}

/// Standalone function outside any impl block.
pub fn standalone_function(param: &str) -> usize {
    param.len()
}

/// Async standalone function.
pub async fn async_standalone(url: &str) -> HashMap<String, String> {
    tokio::time::sleep(std::time::Duration::from_millis(100)).await;
    let mut result = HashMap::new();
    result.insert("url".to_string(), url.to_string());
    result
}

/// Simple macro definition.
#[macro_export]
macro_rules! create_calculator {
    ($value:expr) => {
        SimpleCalculator::new($value)
    };
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_calculator() {
        let mut calc = SimpleCalculator::new(0);
        assert_eq!(calc.add(10), 10);
        assert_eq!(calc.multiply(2), 20);
    }

    #[test]
    fn test_square() {
        assert_eq!(SimpleCalculator::square(5), 25);
    }
}
